cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -414,40 +430,27 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const FileSystemUri& f
     return std::make_shared<internal::MockFileSystem>(internal::CurrentTimePoint());
   }
 
-  // TODO add support for S3 URIs
   return Status::Invalid("Unrecognized filesystem type in URI: ", uri_string);
 }
 
 }  // namespace
 
 Result<std::shared_ptr<FileSystem>> FileSystemFromUri(const std::string& uri_string,
                                                       std::string* out_path) {
-  Status st;
-  FileSystemUri fsuri;
-  auto maybe_fsuri = ParseFileSystemUri(uri_string);
-  if (maybe_fsuri.ok()) {
-    fsuri = *std::move(maybe_fsuri);
-  } else {
-    st = maybe_fsuri.status();
-#ifdef _WIN32
-    // The user may be passing a local path with backslashes, which fails
-    // URI parsing.  Try again with forward slashes, but make sure the
-    // resulting URI points to a local path.
-    maybe_fsuri = ParseFileSystemUri(internal::ToSlashes(uri_string));
-    if (!maybe_fsuri.ok()) {
-      return st;
-    }
-    fsuri = *std::move(maybe_fsuri);
-    if (!fsuri.is_local) {
-      return st;
-    }
-#else
-    return st;
-#endif
-  }
+  ARROW_ASSIGN_OR_RAISE(auto fsuri, ParseFileSystemUri(uri_string));
   return FileSystemFromUriReal(fsuri, uri_string, out_path);
 }
 
+Result<std::shared_ptr<FileSystem>> FileSystemFromUriOrPath(const std::string& uri_string,
+                                                            std::string* out_path) {
+  ARROW_ASSIGN_OR_RAISE(auto fsuri, ParseFileSystemUriOrPath(uri_string));
```

Comment:
Inlining `ParseFileSystem*` might increase clarity

Reply:
I'm not really convinced.